### DISCUSSION TITLE
Python.dict() is missing since version 24.2. Need help!!

### DISCUSSION THREAD

**Discussion by johnsoez4** [modular/modular#2201] [Q&A]
How can I create in Mojo a Python dictionary object that I can pass as an argument to a custom Python function?
```
# This code example given by kapa.ai no longer works.
from python import Python  
  
fn use_dict() raises:  
 var dictionary = Python.dict()  
 dictionary["fruit"] = "apple"  
 dictionary["starch"] = "potato"  
 print("Fruit: ", dictionary["fruit"])
```
When I pass a Mojo dictionary to my Python function, I get an error.
`invalid call to '__call__': argument #1 cannot be converted from 'Dict[String, String]' to 'PythonObject'`

```
# Mojo code
fn getProjects(inout self) raises -> String:
        var _http = Python.import_module("my_http")
        var url = self.REQUEST_URL + self.PROJECT_ROUTE_URL + "&workspace=" + self._credential.workspace_id
        var headers: Dict[String, String]
        headers["Authorization"] = "Bearer " + self._credential.pat

        # Send the HTTP message.
        var api_message = self._http.ApiMessage("get")
        api_message.append(url)
        api_message.append(headers) # invalid call to '__call__': argument #1 cannot be converted from 'Dict[String, String]' to 'PythonObject'
```
And when I try to use a Python dictionary created from a custom class, Mojo flags an error: 
```expression must be mutable in assignment```

```# Alternative (doesn't work)
var headers: PythonObject = _http.Headers()
headers["Authorization"] = "Bearer " + self._credential.pat # expression must be mutable in assignment
```

**Comment by johnsoez4** [comment]
So I ended up adding a class method to add dictionary items as shown below. This works, but to me it's non-intuitive. 
**[Commentary]** It would be better to use directly an actual Python dictionary to support interoperability with existing  Python libraries and modules without having to resort to a wrapper or special method. **How many existing/new libraries will never be ported to Mojo, and so will require this approach to  use them in a Mojo application?**
```
# Added a class method to add dictionary items.
fn getProjects(inout self) raises -> String:
        var _http = Python.import_module("my_http")
        var url = self.REQUEST_URL + self.PROJECT_ROUTE_URL + "&workspace=" + self._credential.workspace_id

        var headers: PythonObject = _http.Headers() # A Python dictionary class
        headers.add_kv("Authorization", "Bearer " + self._credential.pat) # Add dictionary items
```

**Comment by Brian-M-J** [comment]
> And when I try to use a Python dictionary created from a custom class, Mojo flags an error:
`expression must be mutable in assignment`

For me, directly using `__setitem__` gets around this issue:
```mojo
from python import Python


fn use_dict() raises:
    var dictionary = Python.evaluate("dict()")  # Python.evaluate("{}") also works
    dictionary.__setitem__("fruit", "apple")
    dictionary.__setitem__("starch", "potato")
    print("Fruit: ", dictionary["fruit"])  # Prints "Fruit:  apple"


fn main() raises:
    use_dict()

```

**Reply by johnsoez4** [reply]
I cannot find where, but someone suggested this code which also works.
`PythonObject(Dict[PythonObject, PythonObject]())`

**Reply by johnsoez4** [reply]
Found it in the Mojo changelog for v24.3 (2024-05-02).
https://docs.modular.com/mojo/changelog
![image](https://github.com/modularml/mojo/assets/133253549/2e82c25c-f862-4fa5-8d87-89e2956e9193)

**Comment by danka12324** [comment]
create a Python dictionary object in Mojo, which one can be entered as an argument for Python users?

Sample code provided by kapa.ai It doesn't work anymore.


importing Python from python

def use_dict():
calling the NotImplementedError error
var dictionary = Python.dict()
dictionary["fruits"] = "apple"
dictionary["starch"] = "potato"
print("Fruit:", dictionary["fruit"])
When I start merging with Mojo in mojo python, digitization occurs.


invalid call to '__call__': argument #1 cannot be converted from 'Dict[string, String]' to 'PythonObject'
Mojo code:


def getProjects(without self) calls -> String:
var _http =Python.import_module("my_http")
var url =self.REQUEST_URL + self.PROJECT_ROUTE_URL + "&workspace=" + self._credential.workspace_id
variable headers: Dict[String, String]
 headers["Authorization"] ="Bearer" + self._credential.pat

 # Cancel the HTTP connection.
 var api_message = self._http.ApiMessage("get")
api_message.append(url
